pkg/xds/sync/dataplane_sync_tracker.go

```diff
@@ -17,66 +17,94 @@ var (
+// that indicates that the stream was already associated
 type dataplaneSyncTracker struct {
 	newDataplaneWatchdog NewDataplaneWatchdogFunc
 
-	mu      stdsync.RWMutex // protects access to the fields below
-	streams map[int64]context.CancelFunc
+	stdsync.RWMutex    // protects access to the fields below
+	streamsAssociation map[core_xds.StreamID]core_model.ResourceKey
+	dpStreams          map[core_model.ResourceKey]streams
 }
 
 // OnStreamOpen is called once an xDS stream is open with a stream ID and the type URL (or "" for ADS).
 // Returning an error will end processing and close the stream. OnStreamClosed will still be called.
-func (t *dataplaneSyncTracker) OnStreamOpen(ctx context.Context, streamID int64, typ string) error {
+func (t *dataplaneSyncTracker) OnStreamOpen(ctx context.Context, streamID core_xds.StreamID, typ string) error {
 	return nil
 }
 
 // OnStreamClosed is called immediately prior to closing an xDS stream with a stream ID.
-func (t *dataplaneSyncTracker) OnStreamClosed(streamID int64) {
-	t.mu.Lock() // write access to the map of all ADS streams
-	defer t.mu.Unlock()
-
-	defer delete(t.streams, streamID)
-
-	if stopWatchdog := t.streams[streamID]; stopWatchdog != nil {
-		stopWatchdog()
+func (t *dataplaneSyncTracker) OnStreamClosed(streamID core_xds.StreamID) {
+	t.Lock()
+	defer t.Unlock()
+
+	dp, hasAssociation := t.streamsAssociation[streamID]
+	delete(t.streamsAssociation, streamID)
+
+	if hasAssociation {
+		streams := t.dpStreams[dp]
+		delete(streams.activeStreams, streamID)
+		if len(streams.activeStreams) == 0 { // no stream is active, cancel watchdog
+			if streams.watchdogCancel != nil {
+				streams.watchdogCancel()
+			}
+			delete(t.dpStreams, dp)
+		}
 	}
 }
 
 // OnStreamRequest is called once a request is received on a stream.
 // Returning an error will end processing and close the stream. OnStreamClosed will still be called.
-func (t *dataplaneSyncTracker) OnStreamRequest(streamID int64, req *envoy.DiscoveryRequest) error {
-	t.mu.RLock() // read access to the map of all ADS streams
-	_, hasWatchdog := t.streams[streamID]
-	t.mu.RUnlock()
+func (t *dataplaneSyncTracker) OnStreamRequest(streamID core_xds.StreamID, req *envoy.DiscoveryRequest) error {
+	t.RLock()
+	_, alreadyAssociated := t.streamsAssociation[streamID]
+	t.RUnlock()
 
-	if hasWatchdog {
+	if alreadyAssociated {
 		return nil
 	}
 
-	t.mu.Lock() // write access to the map of all ADS streams
-	defer t.mu.Unlock()
-
 	if id, err := core_xds.ParseProxyId(req.Node); err == nil {
 		dataplaneKey := core_model.ResourceKey{Mesh: id.Mesh, Name: id.Name}
-
-		// kick off watchdag for that Dataplane
-		stopCh := make(chan struct{})
-		t.streams[streamID] = context.CancelFunc(func() {
-			close(stopCh)
-		})
-		go t.newDataplaneWatchdog(dataplaneKey, streamID).Start(stopCh)
-		dataplaneSyncTrackerLog.V(1).Info("started Watchdog for a Dataplane", "streamid", streamID, "proxyId", id, "dataplaneKey", dataplaneKey)
+		t.Lock()
```

Comment:
Is it okay that checking of existence in line 81 is not atomic with the actual updating of streams? So after the second lock, some associated resource might exist

Reply:
`OnStreamRequest` for given `streamID` is synchronous so if I understand your question it should be fine